[MLIR][NVVM] Add NVVMRequiresSM op traits

Wolfram70 opened the pull request:
Motivation:
Currently, the NVVMOps are not verified against the supported SM architectures. This can manifest as an ISel failure in the NVPTX LLVM backend during CodeGen to PTX ISA. This PR addresses this issue by adding verifier checks for Target-SM architectures in the NVVM Dialect itself, thereby catching the errors early on.

Summary:
* Parametric traits named `NVVMRequiresSM` and `NVVMRequiresSMa` are added to facilitate the version checks for typical and arch-accelerated versions respectively.
* These traits can be attached to any NVVM Op to enable the checks for the particular Op. (example shown below)
* An attribute interface called named `TargetAttrVerifyInterface` is added to the GPU dialect which any target attribute seeking to perform target-verification on the module can implement.
* The checks are performed by the `NVVMTargetAttr` (implementing the `TargetAttrVerifyInterface` interface) when called from the GPU module verifier where it walks through the module and performs the checks for Ops with the `NVVMRequiresSM` traits.
* A few Ops in `NVVMOps.td` have been updated to serve as examples.

Example Usage:
```
       def NVVM_ReduxOp : NVVM_Op<"redux.sync"> {...} 
 ----> def NVVM_ReduxOp : NVVM_Op<"redux.sync", [NVVMRequiresSM<80>]> {...}

       def NVVM_WgmmaFenceAlignedOp : NVVM_Op<"wgmma.fence.aligned"> {...}
 ----> def NVVM_WgmmaFenceAlignedOp : NVVM_Op<"wgmma.fence.aligned", [NVVMRequiresSMa<[90]>]> {...}
```

schwarzschild-radius commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.h:
```diff
@@ -0,0 +1,91 @@
+  const StringRef getArchString() const { return ArchString; }
+
+  void parse(StringRef SMVersion) {
+    ArchAccelerated = (SMVersion[SMVersion.size() - 1] == 'a');
```
Maybe replace `SMVersion[SMVersion.size()]` with `SMVersion.back()`?

durga4github commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.h:
```diff
@@ -0,0 +1,91 @@
+
+  const StringRef getArchString() const { return ArchString; }
+
+  void parse(StringRef SMVersion) {
```
Can you add a few lines of comment on the input/output of this parse function?

Wolfram70 replied:
Added a description in the latest revision, thanks!

joker-eph commented:
Can you add the motivation and usage in the description of the PR? Right now this describe what it does but not really provide context on why.

> The check can be disabled with the existing --mlir-very-unsafe-disable-verifier-on-parsing flag (tests added in nvvm-check-targetSM-disabled.mlir)

Please remove this: 1) as the name indicates, this isn't something we'd want anyone to use, and 2) this is misleading: it does not disable any check other than on initial parsing.

joker-eph commented on mlir/lib/Dialect/LLVMIR/IR/NVVMDialect.cpp:
```diff
@@ -1439,6 +1440,36 @@ NVVMTargetAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+llvm::SmallVector<NVVMCheckSMVersion> NVVM::getTargetSMVersions(Operation *op) {
+  llvm::SmallVector<NVVMCheckSMVersion> TargetSMVersions;
+  gpu::GPUModuleOp GPUModule = op->getParentOfType<gpu::GPUModuleOp>();
+  if (GPUModule && GPUModule->hasAttr("targets")) {
+    ArrayAttr Targets = dyn_cast<ArrayAttr>(GPUModule->getAttr("targets"));
```
You have a dyn_cast without checking here, that could crash on the next line.

```suggestion
  if (!GPUModule) return {};
  GPUModule->hasAttr("targets")) {
  auto targets = GPUModule->getAttrOfType<ArrayAttr>("targets");
  if (!target) return {};
```

joker-eph commented on mlir/lib/Dialect/LLVMIR/IR/NVVMDialect.cpp:
```diff
@@ -1439,6 +1440,36 @@ NVVMTargetAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+//===----------------------------------------------------------------------===//
+llvm::SmallVector<NVVMCheckSMVersion> NVVM::getTargetSMVersions(Operation *op) {
+  llvm::SmallVector<NVVMCheckSMVersion> TargetSMVersions;
+  gpu::GPUModuleOp GPUModule = op->getParentOfType<gpu::GPUModuleOp>();
```
That will make it for a fairly costly verifier: for every op implementing this trait we'll walk the IR up and find the module, and then again have to lookup the attributes.

I believe a more efficient implementation would be to hook through an attribute verifier on the GPU module which would lookup the attribute and then walk the entire module once to find operation with the traits.

Wolfram70 replied:
Thanks for the suggestion! I will look into this.

Wolfram70 replied:
I have changed it in the latest revision to perform the check through `NVVMTargetAttr` and walk through the module only once. Could you please take another look to see if it is okay?

Wolfram70 commented on mlir/lib/Dialect/LLVMIR/IR/NVVMDialect.cpp:
```diff
@@ -1439,6 +1440,36 @@ NVVMTargetAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+// Requires minimum target SM trait helper functions
+//===----------------------------------------------------------------------===//
+llvm::SmallVector<NVVMCheckSMVersion> NVVM::getTargetSMVersions(Operation *op) {
+  llvm::SmallVector<NVVMCheckSMVersion> TargetSMVersions;
```
Changed in the latest revision, thanks!

joker-eph commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.h:
```diff
@@ -0,0 +1,88 @@
+struct NVVMCheckSMVersion {
+  int archVersion;
+  bool archAccelerated;
+  std::string archString;
```
This field does not seem used anywhere, can we remove it?

Wolfram70 replied:
Thanks for pointing out! I missed this. Removed the field in the latest revision.

joker-eph commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.h:
```diff
@@ -0,0 +1,88 @@
+    return archVersion <= TargetSM.archVersion;
+  }
+
+  bool operator==(const NVVMCheckSMVersion &Other) const {
```
```suggestion
  bool operator==(const NVVMCheckSMVersion &other) const {
```

Nit: mlir style is lower-case variable

joker-eph commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.h:
```diff
@@ -0,0 +1,88 @@
+        .getAsInteger(10, archVersion);
+  }
+
+  bool isCompatible(const NVVMCheckSMVersion &TargetSM) const {
```
```suggestion
  bool isCompatible(const NVVMCheckSMVersion &targetSM) const {
```

Nit: mlir style is lower-case variable

joker-eph commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.h:
```diff
@@ -0,0 +1,88 @@
+
+  // Parses the SM version string and sets the archVersion (integer) and
+  // the archAccelerated flag.
+  void parse(StringRef SMVersion) {
```
```suggestion
  void parse(StringRef smVersion) {
```

Nit: mlir style is lower-case variable

joker-eph commented on mlir/lib/Dialect/LLVMIR/IR/NVVMDialect.cpp:
```diff
@@ -1558,6 +1559,25 @@ NVVMTargetAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+  if (!gpuModuleOp)
+    return emitError(gpuModule->getLoc(),
+                     "NVVM target attribute must be attached to a GPU module");
+  gpuModuleOp->walk([&](Operation *op) {
+    if (auto reqOp = llvm::dyn_cast<NVVM::RequiresSMInterface>(op)) {
+      auto requirement = reqOp.getRequiredMinSMVersion();
+      if (!requirement.isCompatible(NVVMCheckSMVersion(getChip()))) {
+        op->emitOpError() << "is not supported on " << getChip();
+        return WalkResult::interrupt();
+      }
+    }
+    return WalkResult::advance();
```
```suggestion
  
  NVVMCheckSMVersion targetVersion(getChip());
  gpuModuleOp->walk([&](Operation *op) {
    if (auto reqOp = llvm::dyn_cast<NVVM::RequiresSMInterface>(op)) {
      NVVMCheckSMVersion requirement = reqOp.getRequiredMinSMVersion();
      if (!requirement.isCompatible(smVersion)) {
        op->emitOpError() << "is not supported on " << getChip();
        return WalkResult::interrupt();
      }
    }
    return WalkResult::advance();
```

There is string parsing involved here, let's do it once instead of for every op in the module.

Wolfram70 replied:
Updated in the latest revision, thanks!

joker-eph reviewed (commented):
This LGTM, with some comments.

Wolfram70 committed:
[MLIR][NVVM] Add NVVMRequiresSM op trait

This change adds the NVVMRequiresSM op trait to the NVVM dialect to
allow tagging NVVM Ops with a minimum required SM version. When a
target SM is able to be determined (through NVVMTargetAttr), this
allows the verification of SM compatibility with the Op without needing
to unnecessarily lower any further down.

joker-eph commented:
LG, but I'd like to see an approval from either @durga4github or @grypp here. Thanks!

durga4github commented:
> LG, but I'd like to see an approval from either @durga4github or @grypp here. Thanks!

I am working closely with the author on this ;-)
Hence, we wanted to get reviews from you and Guray!

> 
> Also please make sure the description reflects the latest changes.

durga4github commented on mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:
```diff
@@ -1793,6 +1793,22 @@ void GPUModuleOp::setTargets(ArrayRef<TargetAttrInterface> targets) {
+
+  if (!targets)
+    return success();
+
```
It would be useful to have an option to disable these SM version checks during development.
Can we add a flag/knob to control this behaviour?

joker-eph replied:
We can't: the IR verifier is absolute, there is no "conditional validity".

That's what made me a bit hesitant about this feature implemented in the verifier, an alternative would be to have a different pass doing this kind of check and run it before translation to LLVM.

Wolfram70 replied:
Could we have a parameter in `NVVMTargetAttr` like a `verify` flag that can be set to disable these checks? In that case, we could return before we walk through the GPU module.

joker-eph replied:
We can, but such feature makes me question that this checks belongs to the IR verifier in the first place.

Wolfram70 replied:
I have added a `verifyTarget` parameter to `NVVMTargetAttr` in the latest revision as a new commit. Please take a look, thanks!

Wolfram70 replied:
> What makes this invariant different from all the other invariants implemented in the verifier?

From what I understand, this is an invariant since if the usage violates it, we will definitely run into errors further down the line during CodeGen. We could have an option to disable it during development to save some compile-time. Perhaps it could be renamed to something more appropriate to indicate that it should be used cautiously? Please let me know if I have misunderstood something or if my reasoning is incorrect.

joker-eph replied:
I don't think that "compile time"  is a valid reason to remove this during development: the impact in terms of compile-time should be low (otherwise I question whether it is suitable for a verifier check in the first place: the verifier runs in production where the compile-time is critical, contrary to the development flow).

I suspect @durga4github didn't have compile-time in mind as a motivation though.

durga4github replied:
yes, I was not thinking about compile-time here.

I would like to have an option to keep these checks "out-of-the-way" during certain development flows.
Especially when we build/test the lowering-upto-llvmir only and not further below.

joker-eph replied:
Isn't the way to keep these checks out-of-the-way just to disable the verifier?

durga4github replied:
Yes, we could do that. But I was looking for a lightweight option, if possible.
i.e. Keep the verifier for the NVVM Dialect Ops (for functionality) but only bypass this version check (to lower to llvmir etc.). When an Op fails in Verifier (until now, before this change), the lowering-to-llvmir itself will most likely fail (since an intrinsic may not exist for that invalid combination). With this, there can be a valid intrinsic to lower, but it may fail due to an invalid SMArch attached.

So, I was unsure if adding a pass solely for the SMVersion checks is a good approach.
But, if it is indeed the right/idiomatic way to achieve this, we will consider that option too.

grypp replied:
The LLVM backend already has those checks, so what we’re doing here is useful but redundant.

How about we take a big kernel and see how much compile-time it takes?

I’m not sure about a pass that verifies the operations. But it’s definitely a good idea to have a way to enable or disable it.

Wolfram70 committed:
Add verifyTarget parameter to NVVMTargetAttr

grypp commented on mlir/include/mlir/Dialect/LLVMIR/NVVMOps.td:
```diff
@@ -167,14 +170,14 @@ class NVVM_SpecialRangeableRegisterOp<string mnemonic> :
 def NVVM_WarpSizeOp : NVVM_SpecialRangeableRegisterOp<"read.ptx.sreg.warpsize">;
 def NVVM_WarpIdOp : NVVM_SpecialRangeableRegisterOp<"read.ptx.sreg.warpid">;
-def NVVM_WarpDimOp : NVVM_SpecialRangeableRegisterOp<"read.ptx.sreg.nwarpid">;
+def NVVM_WarpDimOp : NVVM_SpecialRangeableRegisterOp<"read.ptx.sreg.nwarpid", [NVVMRequiresSM<20>]>;
```
What is the minimum SM that llvm supports? This could be the default attribute, so we don't have to set sm20 everywhere

Wolfram70 replied:
It looks like `sm_20` is the minimum version for the Ops, so made that as the default in the `verifyTarget` function and removed it from the Ops. Thanks!

grypp commented on mlir/include/mlir/Dialect/LLVMIR/NVVMOps.td:
```diff
@@ -2358,8 +2362,8 @@ def NVVM_WgmmaFenceAlignedOp : NVVM_Op<"wgmma.fence.aligned"> {
-def NVVM_WgmmaGroupSyncAlignedOp : NVVM_Op<"wgmma.commit.group.sync.aligned">,
-  Arguments<(ins )> {
+def NVVM_WgmmaGroupSyncAlignedOp : NVVM_Op<"wgmma.commit.group.sync.aligned",  
+                              [NVVMRequiresSM<90, /*ArchAccelerated*/"true">]> {
```
Should we mark at least all sm90? So we have at least a good test coverage

grypp commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.h:
```diff
@@ -0,0 +1,79 @@
+
+struct NVVMCheckSMVersion {
+  int archVersion;
+  bool archAccelerated;
```
Let's add a comment to explain ehat is this about

Wolfram70 replied:
Oh sorry! I thought it was for the whole struct and added a comment there. I've now added comments to explain the individual fields, thanks!

Wolfram70 committed:
address comments

Wolfram70 committed:
address comments

Wolfram70 committed:
fix formatting

grypp commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.h:
```diff
@@ -0,0 +1,89 @@
+struct NVVMCheckSMVersion {
+  int archVersion;
+  bool archAccelerated;
+  bool exactMatch;
```
again add doc-string, what does `exactMatch` means

Wolfram70 replied:
Added a comment to explain it, thanks!

grypp commented on mlir/lib/Dialect/LLVMIR/IR/NVVMDialect.cpp:
```diff
@@ -1536,7 +1537,7 @@ LogicalResult
                        int optLevel, StringRef triple, StringRef chip,
                        StringRef features, DictionaryAttr flags,
-                       ArrayAttr files) {
+                       ArrayAttr files, bool verifyTarget) {
```
where `verifyTarget` is used?

Wolfram70 replied:
This flag is used through `getVerifyTarget()` in the `NVVMTargetAttr::verifyTarget()` method. I've added it here as a parameter to match the signature of the `verify` method.

grypp commented on mlir/lib/Dialect/LLVMIR/IR/NVVMDialect.cpp:
```diff
@@ -1558,6 +1559,34 @@ NVVMTargetAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+
+  auto gpuModuleOp = llvm::dyn_cast<gpu::GPUModuleOp>(gpuModule);
+  if (!gpuModuleOp)
+    return emitError(gpuModule->getLoc(),
+                     "NVVM target attribute must be attached to a GPU module");
```
nit: I'd add {} as multiline

grypp commented on mlir/test/lib/Dialect/Test/TestOps.td:
```diff
@@ -2728,6 +2729,28 @@ def TestLinalgFillOp :
+}
+
+def TestNVVMRequiresAtleastSMArchCondOp : 
+    TEST_Op<"nvvm_requires_sm_atleast_90_aa", [NVVMRequiresSM<90, "true">]> {
```
Why do we create test ops instead of using existing NVVM ops? I might be missing something.

Wolfram70 replied:
I did this mainly to decouple the testing of the trait from the usage in the Ops. My thought was that this would allow testing the different aspects of the trait without having to find the Ops with the exact requirements we need for testing.

grypp commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.h:
```diff
@@ -0,0 +1,89 @@
+  bool archAccelerated;
+  bool exactMatch;
+
+  NVVMCheckSMVersion()
```
We also need `f` as well other than `a` right?

Wolfram70 replied:
Yes, I am aware of the family specific conditionals introduced in CUDA 12.9 recently. I plan to extend the trait to check for this too in a subsequent PR to this one.

grypp reviewed (commented):
I've added more comments. Let's address them and also rebase the PR

Wolfram70 committed:
address comments

Wolfram70 committed:
[MLIR][NVVM] Add NVVMRequiresSM op trait

This change adds the NVVMRequiresSM op trait to the NVVM dialect to
allow tagging NVVM Ops with a minimum required SM version. When a
target SM is able to be determined (through NVVMTargetAttr), this
allows the verification of SM compatibility with the Op without needing
to unnecessarily lower any further down.

Wolfram70 committed:
fix formatting

grypp commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.h:
```diff
@@ -0,0 +1,96 @@
+  // true if the SM version is accelerated (Ex. sm_90a vs sm_90)
+  bool archAccelerated;
+
+  // true if the target SM version must exactly match this one
+  // (both archVersion and archAccelerated)
+  // Ex. sm_90a with exactMatch = false will also match with
+  // sm_100a, sm_120a, etc...
```
```suggestion
  // Set to true if the target SM version must match exactly
  // (both archVersion and archAccelerated).
  // For example, sm_90a with exactMatch = false will also match
  // sm_100a, sm_120a, etc.
```

grypp commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.h:
```diff
@@ -0,0 +1,96 @@
+      : archVersion(archVersion), archAccelerated(archAccelerated),
+        exactMatch(exactMatch) {}
+
+  // Parses the SM version string and sets the archVersion (integer) and
+  // the archAccelerated flag.
```
```suggestion
  // Parses the SM version string and sets the archVersion (as an integer)
  // and the archAccelerated flag.
```

grypp commented on mlir/include/mlir/Dialect/GPU/IR/CompilationAttrInterfaces.td:
```diff
@@ -55,6 +55,20 @@ def GPUTargetAttrInterface : AttrInterface<"TargetAttrInterface"> {
 
+def GPUTargetAttrVerifyInterface : AttrInterface<"TargetAttrVerifyInterface"> {
+  let description = [{
+    Interface for GPU target attributes that need to verify the target attribute
+    for the given GPU module.
```
```suggestion
    Interface for GPU target attributes that verify the target attribute
    of a given GPU module.
```

grypp commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.h:
```diff
@@ -0,0 +1,96 @@
+struct NVVMCheckSMVersion {
+  int archVersion;
+
+  // true if the SM version is accelerated (Ex. sm_90a vs sm_90)
```
```suggestion
  // Set to true if the SM version is accelerated (e.g., sm_90a).
```

Wolfram70 committed:
Update comments in NVVMTraits.h

Apply suggestions from code review

Co-authored-by: Guray Ozen <guray.ozen@gmail.com>

Wolfram70 committed:
Update GPUTargetVerifyInterface description

Co-authored-by: Guray Ozen <guray.ozen@gmail.com>

grypp commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.td:
```diff
@@ -0,0 +1,38 @@
+                    !cast<string>(minVersion) # "," # isArchAccelerated # ","
+                      # exactMatch>;
+                      
+def NVVMRequiresSM90a : NVVMRequiresSM<90, "true", "true">;
```
without `a` one can just use this: 
```
NVVMRequiresSM<90> 
```

but if there is `a` we define a shortcut that's not intuitive to find out. 
```
NVVMRequiresSM90a
``` 

Can we have this?
```
NVVMRequiresSMa<90> 
```

Wolfram70 replied:
Oh yes, that looks better. Changed it in the latest revision. Thanks!

Wolfram70 committed:
address comments

Wolfram70 commented on mlir/include/mlir/Dialect/LLVMIR/NVVMTraits.td:
```diff
@@ -0,0 +1,38 @@
+//===-- NVVMTraits.td - NVVM Traits ------------------------*- tablegen -*-===//
```
Renamed the files in the latest revision.

Wolfram70 committed:
add NVVMRequiresSMa for arch-accelerated SM versions

Wolfram70 committed:
fix formatting

grypp commented on mlir/include/mlir/Dialect/LLVMIR/NVVMRequiresSMTraits.td:
```diff
@@ -27,12 +27,17 @@ def RequiresSMInterface: OpInterface<"RequiresSMInterface"> {
-                      # exactMatch>;
-                      
-class NVVMRequiresSMa<int version> : NVVMRequiresSM<version, "true", "true">;
+class NVVMRequiresSM<int minVersion> :
```
```suggestion
// OP requires a specified minimum SM value or higher; 
// it is not architecture-specific.
"OP requires an exact SM match along with architecture acceleration."
class NVVMRequiresSM<int minVersion> :
```

grypp commented on mlir/include/mlir/Dialect/LLVMIR/NVVMRequiresSMTraits.td:
```diff
@@ -27,12 +27,17 @@ def RequiresSMInterface: OpInterface<"RequiresSMInterface"> {
+               !if(!eq(a, ""), b, !if(!eq(b, ""), a, !strconcat(a, sep, b))));
+}
+
+class NVVMRequiresSMa<list<int> smVersions> :
+  ParamNativeOpTrait<"NVVMRequiresSMa",
+                      StrJoin<",", !foreach(vers, smVersions,
+                                            !cast<string>(vers))>.ret>;
```
```suggestion
// OP requires an exact SM match along with 
// architecture acceleration.
class NVVMRequiresSMa<list<int> smVersions> :
  ParamNativeOpTrait<"NVVMRequiresSMa",
                      StrJoin<",", !foreach(vers, smVersions,
                                            !cast<string>(vers))>.ret>;
```

grypp commented on mlir/include/mlir/Dialect/LLVMIR/NVVMOps.td:
```diff
@@ -3136,7 +3136,7 @@ def NVVM_Tcgen05CommitOp : NVVM_Op<"tcgen05.commit", [NVVMRequiresSM<100, "true"
 }
 
-def NVVM_Tcgen05ShiftOp : NVVM_Op<"tcgen05.shift", [NVVMRequiresSM<100, "true", "false">]> {
+def NVVM_Tcgen05ShiftOp : NVVM_Op<"tcgen05.shift", [NVVMRequiresSMa<[100, 101, 103]>]> {
```
Nice!

Wolfram70 committed:
Apply suggestions from code review

Co-authored-by: Guray Ozen <guray.ozen@gmail.com>

grypp commented:
Thanks for bearing with me and handling all the code reviews—I think the PR is ready to land!

Wolfram70 commented:
No worries, thanks a lot!
